presto-main/src/main/java/com/facebook/presto/type/DecimalSaturatedFloorCasts.java

```diff
@@ -95,4 +105,119 @@ private static BigInteger bigintToBigintFloorSaturatedCast(BigInteger value, int
         }
         return unscaledValue;
     }
+
+    public static final SqlScalarFunction DOUBLE_TO_DECIMAL_SATURATED_FLOOR_CAST = SqlScalarFunction.builder(DecimalSaturatedFloorCasts.class)
+            .signature(Signature.builder()
+                    .kind(SCALAR)
+                    .operatorType(SATURATED_FLOOR_CAST)
+                    .argumentTypes(DOUBLE.getTypeSignature())
+                    .returnType(parseTypeSignature("decimal(result_precision,result_scale)", ImmutableSet.of("result_precision", "result_scale")))
+                    .build()
+            )
+            .implementation(b -> b
+                    .methods("doubleToShortDecimal", "doubleToLongDecimal")
+                    .withExtraParameters((context) -> {
+                        int resultPrecision = Ints.checkedCast(context.getLiteral("result_precision"));
+                        int resultScale = Ints.checkedCast(context.getLiteral("result_scale"));
+                        return ImmutableList.of(resultPrecision, resultScale);
+                    })
+            ).build();
+
+    @UsedByGeneratedCode
+    public static long doubleToShortDecimal(double value, int resultPrecision, int resultScale)
+    {
+        return bigDecimalToBigintFloorSaturatedCast(BigDecimal.valueOf(value), resultPrecision, resultScale).longValueExact();
```

Comment:
Fixed by using the `BigDecimal(double)` constructor. With such constructor double can be represented as a `BigDecimal` precisely.